Fixes #24660 - Make EmptyState more flexible

xprazak2 opened the pull request:
Main changes:
* Katello passes a function to `onClick` for primary button on Subscriptions page so we need to support it
* overriding icon type
* new optional param `flattenData` for component mounter to make sure props passed from erb are on correct level, default behaviour remains the same.
* partition tables welcome page now uses the component as a live example how to use it

Commit:
Fixes #24660 - Make EmptyState more flexible

xprazak2 commented:
Current look:

![](https://i.imgur.com/idnyaHA.png)

ohadlevy commented:
/cc @ripcurld

boaz0 commented on webpack/assets/javascripts/react_app/components/common/EmptyState/DefaultEmptyState.js:
```diff
@@ -32,11 +31,12 @@ const DefaultEmptyState = (props) => {
       header={header}
       description={description}
       documentation={documentation ? documentationBlock(documentation) : null}
-      action={<PrimaryActionButton action={action} />}
```
The fact that I added PrimaryActionButton and SecondaryActionButtons here is because that's the default action buttons that are used in Foreman.

However, in Katello the action buttons (both Primary and Secondary) are using Links that wrap a button.
Thus, I introduced `EmptyStatePattern` that allow any plugin to use it in order to create their own Empty State component.

boaz0 reviewed (commented):
This change looks really good.

Just I am in favor to keep `DefaultEmptyState` and `EmptyStatePattern` as they were.

Can you explain why you think it's necessary, though?
Thanks.

amirfefer commented on app/views/ptables/welcome.html.erb:
```diff
@@ -1,12 +1,15 @@
+                            :documentation => { :url => documentation_url("4.4.4PartitionTables") },
+                            :action => { :title => _("Create Partition Table"), :url => new_ptable_path }
+                          }.to_json,
+                          true
```
`mount_react_component` becomes too overloaded, at first look it would be hard to tell what this `true` is for.
how about using hash instead?

xprazak2 commented:
The main motivation here is to remove EmptyState component from Katello and use the one we now have in Foreman instead, which is a driver for these changes. I reverted the buttons back, I moved them because the result was better readable for me and I did not know about the original intent. Now that I do, it actually raises more questions.

> However, in Katello the action buttons (both Primary and Secondary) are using Links that wrap a button.

On what page is that? Subscriptions page uses regular button and manifest history has no primary action. Why do buttons need to be wrapped in links?

![](https://i.imgur.com/ppd8jCe.png)


> Thus, I introduced EmptyStatePattern that allow any plugin to use it in order to create their own Empty State component.

Why would any plugin need that? The look should always be the same, the plugin only needs to specify things to display like title, icon, docs link and what should happen when a button is clicked - that is something our component should be able to handle.

Even if we know about the cases where plugins need their own buttons, shouldn't we support that as well rather than having plugins implement a new version of the same component just because they need to change a button?

boaz0 commented:
@xprazak2 
> On what page is that? Subscriptions page uses regular button and manifest history has no primary action. Why do buttons need to be wrapped in links?

In the subscription page they use EmptyState in Table and the EmptyState itself is using [`LinkContainer`](https://github.com/Katello/katello/blob/master/webpack/move_to_foreman/components/common/EmptyState/index.js#L35-L39)

That being said, if your changes doesn't break Katello, then you got a green light from me.

xprazak2 commented:
I double-checked, the katello pages were working for me.

amirfefer commented on webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePattern.js:
```diff
@@ -4,9 +4,10 @@ import { emptyStatePatternPropTypes } from './EmptyStatePropTypes';
 
 const EmptyStatePattern = (props) => {
   const { documentation, action, secondaryActions } = props;
```
nitpick: add 'iconType', 'header' and 'description' to object destructuring.

amirfefer commented on webpack/assets/javascripts/react_app/components/common/EmptyState/DefaultEmptyState.js:
```diff
@@ -32,6 +33,7 @@ const DefaultEmptyState = (props) => {
   return (
     <EmptyStatePattern
       icon={icon}
+      iconType={iconType || 'pf'}
```
you can use default value through proptype 

```javascript
DefaultEmptyState.defaultProps = {
  iconType: 'pf'
};
```

ohadlevy commented:
whats the status of this pr? thanks!

boaz0 commented:
[test katello]
[test foreman]

ohadlevy commented:
@tstrachota can you review please? thanks

tstrachota commented:
I'm sorry @xprazak2 and @ohadlevy. I wanted to review this since I worked with empty state in Katello, but I didn't find time for that before my PTO. I'm going to be away for next two weeks (back on Oct 22).

I removed myself from the assignees. If somebody else can take a look, please do so.

mmoll commented:
@xprazak2 please check the test failures.

tstrachota commented:
[test foreman]

tstrachota reviewed (dismissed):
Works nicely. I tested this changes in Katello (a PR will follow) and it works seamlessly!

Approved, pending successful test run.

tstrachota commented:
@xprazak2 I didn't spot you've already opened https://github.com/Katello/katello/pull/7637 Sorry for that.
I found out some more changes are necessary in order to make this PR correctly working with Katello. See my proposal in: https://github.com/Katello/katello/pull/7798

tstrachota commented on app/helpers/reactjs_helper.rb:
```diff
@@ -1,7 +1,7 @@
+  def mount_react_component(name, selector, data = [], opts = {})
     javascript_tag defer: 'defer' do
-      "$(tfm.reactMounter.mount('#{name}', '#{selector}', #{data}));".html_safe
+      "$(tfm.reactMounter.mount('#{name}', '#{selector}', #{data}, #{opts[:flatten_data]}));".html_safe
```
This might be causing issues with the tests. Adding `opts[:flatten_data] ||= false` could fix that.

xprazak2 commented:
[test foreman]

tstrachota commented:
[test foreman]

tstrachota commented:
Tests are green, merging. Thank you @xprazak2 ! It's great to see such effort to unify the codebase.
